synapse/storage/databases/main/deviceinbox.py

```diff
@@ -153,12 +156,21 @@ def get_new_messages_for_device_txn(txn):
             txn.execute(
                 sql, (user_id, device_id, last_stream_id, current_stream_id, limit)
             )
+
             messages = []
+            stream_pos = current_stream_id
+
             for row in txn:
                 stream_pos = row[0]
                 messages.append(db_to_json(row[1]))
+
+            # If we don't end up hitting the limit, we still want to return the equivalent
+            # value of `current_stream_id` to the calling function. This is needed as we'll
+            # be processing up to `current_stream_id`, without necessarily fetching a message
+            # with a stream token of `current_stream_id`.
```

Comment:
I'm not really following this, wouldn't this cause us to "skip" rows or something?

Reply:
We shouldn't, though I see that the explanation isn't clear.

For example, let's say `user_id` is `@alice:localhost`, `last_stream_id` is 1, and `current_stream_id` is 5, limit is 3, and this is what the DB looks like:

```
      user_id      |                  device_id                  | stream_id |     message_json   | instance_name 
-------------------+---------------------------------------------+-----------+--------------------+---------------
 @alice:localhost  | HAKDUDVZXZ                                  |         1 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         2 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         3 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         4 | {"content":{"..."} | master
 @bob:localhost    | TYCGLKNPVJ                                  |         5 | {"content":{"..."} | master
(5 rows)
```

We're going to pull out to-device messages with stream ids 2, 3, 4. `len(messages) == limit`, so `stream_pos` is 4. We return `4` alongside the messages, and `4` is stored.

Later on, a new to-device message comes in, so the DB now looks like:

```
      user_id      |                  device_id                  | stream_id |     message_json   | instance_name 
-------------------+---------------------------------------------+-----------+--------------------+---------------
 @alice:localhost  | HAKDUDVZXZ                                  |         1 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         2 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         3 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         4 | {"content":{"..."} | master
 @bob:localhost    | TYCGLKNPVJ                                  |         5 | {"content":{"..."} | master
 @bob:localhost    | TYCGLKNPVJ                                  |         6 | {"content":{"..."} | master
(6 rows)
```

And we call this function with `@alice:localhost`, `last_stream_id` as 4, and `current_stream_id` as 6. Limit is still 3.

We end up not pulling out any messages, as none of them are addressed to `@alice:localhost`. However, we still need to return `6` for `stream_pos`, so that the calling function knows where to check from on the next round. If we returned `4`, then we'd duplicate work checking stream ids 5, 6 again.

---

Thinking more about it now... there is an edge case where we end up returning exactly `<limit>` rows. Say you had the below DB table: 

```
      user_id      |                  device_id                  | stream_id |     message_json   | instance_name 
-------------------+---------------------------------------------+-----------+--------------------+---------------
 @alice:localhost  | HAKDUDVZXZ                                  |         1 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         2 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         3 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         4 | {"content":{"..."} | master
 @bob:localhost    | TYCGLKNPVJ                                  |         5 | {"content":{"..."} | master
(5 rows)
```

And you called `get_new_messages_for_device` with user_id `@alice:localhost`, `last_stream_id` as 1, `current_stream_id` as 5, with limit 3.

The function would check over rows 2 - 5 in the DB, and return rows 2, 3, 4. `len(messages) == limit`, so `stream_pos` is that of the last row we pulled out: 4. However, the function did check row 5 already. Next time round the function will be called with `last_stream_id` as 4. So we'll end up doing some duplicate work checking row 5 again.

I think this only causes a slight drop in perf though, rather than going the other way and causing us to miss rows...

Reply:
After chatting about this it seems that the idea is that if we don't hit the limit that means there's no data left (up to the `current_stream_id`) that relates to the requested filtering (by user/device). Thus it is safe to consider up to `current_stream_id` processed (instead considering only up to the last result's stream position).

I think the following simpler comment describes this?

```suggestion
            # If the limit was not reached we know that there's no more data for this
            # user/device pair up to current_stream_id.
```

Reply:
I'd say that's accurate!

Reply:
Fixed in a25c8dd4c55f79a20b5046c32329a4e27af705aa.